File: Predictive Modeling of Transcription Factor Binding Sites Using Deep Learning/scripts/python/tf_binding_prediction.py
"""
Transcription Factor Binding Prediction using Deep Learning and Machine Learning

This script implements multiple models for predicting transcription factor binding sites:
1. CNN (Convolutional Neural Network)
2. LSTM (Long Short-Term Memory)
3. Random Forest
4. XGBoost
5. SVM (Support Vector Machine)

Author: AI-ML Bioinformatics Team
Date: 2024
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    roc_auc_score, confusion_matrix, classification_report, roc_curve, auc
)
from sklearn.preprocessing import LabelEncoder
import xgboost as xgb
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers, models, callbacks
from tensorflow.keras.utils import to_categorical
import warnings
import os
import pickle
import json
from datetime import datetime

warnings.filterwarnings('ignore')

# Set random seeds for reproducibility
np.random.seed(42)
tf.random.set_seed(42)

class DNASequenceEncoder:
    """Encode DNA sequences to numerical representations."""
    
    def __init__(self):
        self.nucleotide_to_int = {'A': 0, 'T': 1, 'G': 2, 'C': 3}
        self.int_to_nucleotide = {0: 'A', 1: 'T', 2: 'G', 3: 'C'}
    
    def one_hot_encode(self, sequences):
        """
        Convert DNA sequences to one-hot encoded format.
        
        Args:
            sequences: List of DNA sequences (strings)
            
        Returns:
            numpy array of shape (n_samples, sequence_length, 4)
        """
        encoded = np.zeros((len(sequences), len(sequences[0]), 4))
        for i, seq in enumerate(sequences):
            for j, nucleotide in enumerate(seq):
                if nucleotide in self.nucleotide_to_int:
                    encoded[i, j, self.nucleotide_to_int[nucleotide]] = 1
        return encoded
    
    def integer_encode(self, sequences):
        """
        Convert DNA sequences to integer encoding.
        
        Args:
            sequences: List of DNA sequences (strings)
            
        Returns:
            numpy array of shape (n_samples, sequence_length)
        """
        encoded = np.zeros((len(sequences), len(sequences[0])), dtype=int)
        for i, seq in enumerate(sequences):
            for j, nucleotide in enumerate(seq):
                if nucleotide in self.nucleotide_to_int:
                    encoded[i, j] = self.nucleotide_to_int[nucleotide]
        return encoded

class TFBindingPredictor:
    """Main class for Transcription Factor Binding Prediction."""
    
    def __init__(self, data_path='data/genomics_data.csv'):
        self.data_path = data_path
        self.encoder = DNASequenceEncoder()
        self.models = {}
        self.history = {}
        self.results = {}
        
    def load_data(self):
        """Load and preprocess the genomics data."""
        print("Loading data...")
        df = pd.read_csv(self.data_path)
        print(f"Data shape: {df.shape}")
        print(f"Label distribution:\n{df['Labels'].value_counts()}")
        
        sequences = df['Sequences'].values
        labels = df['Labels'].values
        
        # Encode sequences
        X_one_hot = self.encoder.one_hot_encode(sequences)
        X_int = self.encoder.integer_encode(sequences)
        
        # Split data
        self.X_train_oh, self.X_test_oh, self.y_train, self.y_test = train_test_split(
            X_one_hot, labels, test_size=0.2, random_state=42, stratify=labels
        )
        self.X_train_int, self.X_test_int, _, _ = train_test_split(
            X_int, labels, test_size=0.2, random_state=42, stratify=labels
        )
        
        # Reshape for traditional ML models (flatten one-hot)
        self.X_train_flat = self.X_train_oh.reshape(self.X_train_oh.shape[0], -1)
        self.X_test_flat = self.X_test_oh.reshape(self.X_test_oh.shape[0], -1)
        
        print(f"Training set size: {self.X_train_oh.shape[0]}")
        print(f"Test set size: {self.X_test_oh.shape[0]}")
        
    def build_cnn_model(self, input_shape, num_classes=1):
        """Build a CNN model for sequence classification."""
        model = models.Sequential([
            layers.Conv1D(64, kernel_size=3, activation='relu', input_shape=input_shape),
            layers.BatchNormalization(),
            layers.MaxPooling1D(2),
            layers.Conv1D(128, kernel_size=3, activation='relu'),
            layers.BatchNormalization(),
            layers.MaxPooling1D(2),
            layers.Conv1D(256, kernel_size=3, activation='relu'),
            layers.BatchNormalization(),
            layers.GlobalMaxPooling1D(),
            layers.Dense(128, activation='relu'),
            layers.Dropout(0.5),
            layers.Dense(64, activation='relu'),
            layers.Dropout(0.3),
            layers.Dense(num_classes, activation='sigmoid')
        ])
        
        model.compile(
            optimizer=keras.optimizers.Adam(learning_rate=0.001),
            loss='binary_crossentropy',
            metrics=['accuracy', 'precision', 'recall']
        )
        
        return model
    
    def build_lstm_model(self, input_shape, num_classes=1):
        """Build an LSTM model for sequence classification."""
        model = models.Sequential([
            layers.LSTM(128, return_sequences=True, input_shape=input_shape),
            layers.BatchNormalization(),
            layers.Dropout(0.3),
            layers.LSTM(64, return_sequences=False),
            layers.BatchNormalization(),
            layers.Dropout(0.3),
            layers.Dense(64, activation='relu'),
            layers.Dropout(0.2),
            layers.Dense(num_classes, activation='sigmoid')
        ])
        
        model.compile(
            optimizer=keras.optimizers.Adam(learning_rate=0.001),
            loss='binary_crossentropy',
            metrics=['accuracy', 'precision', 'recall']
        )
        
        return model
    
    def train_cnn(self, epochs=50, batch_size=32):
        """Train CNN model."""
        print("\nTraining CNN model...")
        input_shape = (self.X_train_oh.shape[1], self.X_train_oh.shape[2])
        model = self.build_cnn_model(input_shape)
        
        early_stopping = callbacks.EarlyStopping(
            monitor='val_loss', patience=10, restore_best_weights=True
        )
        reduce_lr = callbacks.ReduceLROnPlateau(
            monitor='val_loss', factor=0.5, patience=5, min_lr=1e-7
        )
        
        self.history['cnn'] = model.fit(
            self.X_train_oh, self.y_train,
            validation_split=0.2,
            epochs=epochs,
            batch_size=batch_size,
            callbacks=[early_stopping, reduce_lr],
            verbose=1
        )
        
        self.models['cnn'] = model
        return model
    
    def train_lstm(self, epochs=50, batch_size=32):
        """Train LSTM model."""
        print("\nTraining LSTM model...")
        input_shape = (self.X_train_oh.shape[1], self.X_train_oh.shape[2])
        model = self.build_lstm_model(input_shape)
        
        early_stopping = callbacks.EarlyStopping(
            monitor='val_loss', patience=10, restore_best_weights=True
        )
        reduce_lr = callbacks.ReduceLROnPlateau(
            monitor='val_loss', factor=0.5, patience=5, min_lr=1e-7
        )
        
        self.history['lstm'] = model.fit(
            self.X_train_oh, self.y_train,
            validation_split=0.2,
            epochs=epochs,
            batch_size=batch_size,
            callbacks=[early_stopping, reduce_lr],
            verbose=1
        )
        
        self.models['lstm'] = model
        return model
    
    def train_random_forest(self, n_estimators=100, max_depth=20):
        """Train Random Forest model."""
        print("\nTraining Random Forest model...")
        model = RandomForestClassifier(
            n_estimators=n_estimators,
            max_depth=max_depth,
            random_state=42,
            n_jobs=-1,
            class_weight='balanced'
        )
        model.fit(self.X_train_flat, self.y_train)
        self.models['random_forest'] = model
        return model
    
    def train_xgboost(self, n_estimators=100, max_depth=6):
        """Train XGBoost model."""
        print("\nTraining XGBoost model...")
        model = xgb.XGBClassifier(
            n_estimators=n_estimators,
            max_depth=max_depth,
            learning_rate=0.1,
            random_state=42,
            eval_metric='logloss',
            use_label_encoder=False
        )
        model.fit(self.X_train_flat, self.y_train)
        self.models['xgboost'] = model
        return model
    
    def train_svm(self):
        """Train SVM model."""
        print("\nTraining SVM model...")
        # Use a subset for SVM due to computational complexity
        sample_size = min(5000, len(self.X_train_flat))
        indices = np.random.choice(len(self.X_train_flat), sample_size, replace=False)
        X_train_sample = self.X_train_flat[indices]
        y_train_sample = self.y_train[indices]
        
        model = SVC(kernel='rbf', probability=True, random_state=42, class_weight='balanced')
        model.fit(X_train_sample, y_train_sample)
        self.models['svm'] = model
        return model
    
    def evaluate_model(self, model_name, model, X_test, y_test):
        """Evaluate a model and return metrics."""
        if model_name in ['cnn', 'lstm']:
            y_pred_proba = model.predict(X_test, verbose=0).flatten()
            y_pred = (y_pred_proba > 0.5).astype(int)
        else:
            y_pred = model.predict(X_test)
            y_pred_proba = model.predict_proba(X_test)[:, 1]
        
        metrics = {
            'accuracy': accuracy_score(y_test, y_pred),
            'precision': precision_score(y_test, y_pred, zero_division=0),
            'recall': recall_score(y_test, y_pred, zero_division=0),
            'f1': f1_score(y_test, y_pred, zero_division=0),
            'roc_auc': roc_auc_score(y_test, y_pred_proba),
            'confusion_matrix': confusion_matrix(y_test, y_pred).tolist(),
            'y_pred': y_pred.tolist(),
            'y_pred_proba': y_pred_proba.tolist()
        }
        
        return metrics
    
    def evaluate_all_models(self):
        """Evaluate all trained models."""
        print("\nEvaluating all models...")
        
        for model_name, model in self.models.items():
            if model_name in ['cnn', 'lstm']:
                X_test = self.X_test_oh
            else:
                X_test = self.X_test_flat
            
            metrics = self.evaluate_model(model_name, model, X_test, self.y_test)
            self.results[model_name] = metrics
            
            print(f"\n{model_name.upper()} Results:")
            print(f"Accuracy: {metrics['accuracy']:.4f}")
            print(f"Precision: {metrics['precision']:.4f}")
            print(f"Recall: {metrics['recall']:.4f}")
            print(f"F1-Score: {metrics['f1']:.4f}")
            print(f"ROC-AUC: {metrics['roc_auc']:.4f}")
    
    def plot_training_history(self, model_name):
        """Plot training history for deep learning models."""
        if model_name not in self.history:
            return
        
        history = self.history[model_name].history
        fig, axes = plt.subplots(1, 2, figsize=(12, 4))
        
        # Plot accuracy
        axes[0].plot(history['accuracy'], label='Train Accuracy')
        axes[0].plot(history['val_accuracy'], label='Val Accuracy')
        axes[0].set_title(f'{model_name.upper()} - Accuracy')
        axes[0].set_xlabel('Epoch')
        axes[0].set_ylabel('Accuracy')
        axes[0].legend()
        axes[0].grid(True)
        
        # Plot loss
        axes[1].plot(history['loss'], label='Train Loss')
        axes[1].plot(history['val_loss'], label='Val Loss')
        axes[1].set_title(f'{model_name.upper()} - Loss')
        axes[1].set_xlabel('Epoch')
        axes[1].set_ylabel('Loss')
        axes[1].legend()
        axes[1].grid(True)
        
        plt.tight_layout()
        plt.savefig(f'results/{model_name}_training_history.png', dpi=300, bbox_inches='tight')
        plt.close()
    
    def plot_confusion_matrices(self):
        """Plot confusion matrices for all models."""
        n_models = len(self.results)
        fig, axes = plt.subplots(1, n_models, figsize=(5*n_models, 4))
        
        if n_models == 1:
            axes = [axes]
        
        for idx, (model_name, metrics) in enumerate(self.results.items()):
            cm = np.array(metrics['confusion_matrix'])
            sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', ax=axes[idx])
            axes[idx].set_title(f'{model_name.upper()} Confusion Matrix')
            axes[idx].set_ylabel('True Label')
            axes[idx].set_xlabel('Predicted Label')
        
        plt.tight_layout()
        plt.savefig('results/confusion_matrices.png', dpi=300, bbox_inches='tight')
        plt.close()
    
    def plot_roc_curves(self):
        """Plot ROC curves for all models."""
        plt.figure(figsize=(10, 8))
        
        for model_name, metrics in self.results.items():
            fpr, tpr, _ = roc_curve(self.y_test, metrics['y_pred_proba'])
            roc_auc = metrics['roc_auc']
            plt.plot(fpr, tpr, label=f'{model_name.upper()} (AUC = {roc_auc:.3f})')
        
        plt.plot([0, 1], [0, 1], 'k--', label='Random')
        plt.xlim([0.0, 1.0])
        plt.ylim([0.0, 1.05])
        plt.xlabel('False Positive Rate')
        plt.ylabel('True Positive Rate')
        plt.title('ROC Curves for All Models')
        plt.legend(loc='lower right')
        plt.grid(True)
        plt.savefig('results/roc_curves.png', dpi=300, bbox_inches='tight')
        plt.close()
    
    def save_models(self, directory='models'):
        """Save all trained models."""
        os.makedirs(directory, exist_ok=True)
        
        for model_name, model in self.models.items():
            if model_name in ['cnn', 'lstm']:
                model.save(f'{directory}/{model_name}_model.h5')
            else:
                with open(f'{directory}/{model_name}_model.pkl', 'wb') as f:
                    pickle.dump(model, f)
        
        # Save results
        with open(f'{directory}/results.json', 'w') as f:
            json.dump(self.results, f, indent=2)
        
        print(f"\nModels saved to {directory}/")
    
    def predict(self, sequences, model_name='cnn'):
        """Predict TF binding for new sequences."""
        if model_name not in self.models:
            raise ValueError(f"Model {model_name} not found. Available models: {list(self.models.keys())}")
        
        model = self.models[model_name]
        
        if model_name in ['cnn', 'lstm']:
            X = self.encoder.one_hot_encode(sequences)
            predictions = model.predict(X, verbose=0).flatten()
        else:
            X_one_hot = self.encoder.one_hot_encode(sequences)
            X_flat = X_one_hot.reshape(X_one_hot.shape[0], -1)
            predictions = model.predict_proba(X_flat)[:, 1]
        
        return predictions

def main():
    """Main function to run the TF binding prediction pipeline."""
    # Create results directory
    os.makedirs('results', exist_ok=True)
    os.makedirs('models', exist_ok=True)
    
    # Initialize predictor
    predictor = TFBindingPredictor(data_path='data/genomics_data.csv')
    
    # Load data
    predictor.load_data()
    
    # Train models
    predictor.train_cnn(epochs=50, batch_size=32)
    predictor.train_lstm(epochs=50, batch_size=32)
    predictor.train_random_forest(n_estimators=100, max_depth=20)
    predictor.train_xgboost(n_estimators=100, max_depth=6)
    predictor.train_svm()
    
    # Evaluate all models
    predictor.evaluate_all_models()
    
    # Plot results
    predictor.plot_training_history('cnn')
    predictor.plot_training_history('lstm')
    predictor.plot_confusion_matrices()
    predictor.plot_roc_curves()
    
    # Save models
    predictor.save_models()
    
    print("\nPipeline completed successfully!")
    print("Results saved in 'results/' directory")
    print("Models saved in 'models/' directory")

if __name__ == '__main__':
    main()

File: Predictive Modeling of Transcription Factor Binding Sites Using Deep Learning/scripts/r/tf_binding_prediction.R
# Transcription Factor Binding Prediction using Machine Learning in R
#
# This script implements multiple models for predicting transcription factor binding sites:
# 1. Random Forest
# 2. XGBoost
# 3. SVM (Support Vector Machine)
# 4. Neural Network
# 5. Logistic Regression
#
# Author: AI-ML Bioinformatics Team
# Date: 2024

# Load required libraries
suppressPackageStartupMessages({
  library(tidyverse)
  library(caret)
  library(randomForest)
  library(xgboost)
  library(e1071)
  library(neuralnet)
  library(pROC)
  library(ggplot2)
  library(reshape2)
  library(doParallel)
  library(gridExtra)
})

# Set random seed for reproducibility
set.seed(42)

# Enable parallel processing
cl <- makeCluster(detectCores() - 1)
registerDoParallel(cl)

#' Encode DNA sequences to numerical representations
#' 
#' @param sequences Vector of DNA sequences (strings)
#' @return Matrix of one-hot encoded sequences
encode_dna_sequences <- function(sequences) {
  nucleotide_to_int <- c('A' = 1, 'T' = 2, 'G' = 3, 'C' = 4)
  max_len <- max(nchar(sequences))
  n_seqs <- length(sequences)
  
  # Create one-hot encoded matrix
  encoded <- matrix(0, nrow = n_seqs, ncol = max_len * 4)
  
  for (i in 1:n_seqs) {
    seq <- strsplit(sequences[i], "")[[1]]
    for (j in 1:length(seq)) {
      if (seq[j] %in% names(nucleotide_to_int)) {
        idx <- (j - 1) * 4 + nucleotide_to_int[seq[j]]
        encoded[i, idx] <- 1
      }
    }
  }
  
  return(encoded)
}

#' Load and preprocess genomics data
#' 
#' @param data_path Path to the CSV file
#' @return List containing training and test sets
load_and_preprocess_data <- function(data_path = "data/genomics_data.csv") {
  cat("Loading data...\n")
  
  # Load data
  df <- read.csv(data_path, stringsAsFactors = FALSE)
  cat(sprintf("Data shape: %d rows, %d columns\n", nrow(df), ncol(df)))
  cat(sprintf("Label distribution:\n"))
  print(table(df$Labels))
  
  # Encode sequences
  sequences <- df$Sequences
  labels <- as.factor(df$Labels)
  
  cat("Encoding DNA sequences...\n")
  X_encoded <- encode_dna_sequences(sequences)
  
  # Create data frame
  data_df <- as.data.frame(X_encoded)
  data_df$Label <- labels
  
  # Split data
  train_index <- createDataPartition(data_df$Label, p = 0.8, list = FALSE)
  train_data <- data_df[train_index, ]
  test_data <- data_df[-train_index, ]
  
  X_train <- train_data[, -ncol(train_data)]
  y_train <- train_data$Label
  X_test <- test_data[, -ncol(test_data)]
  y_test <- test_data$Label
  
  cat(sprintf("Training set size: %d\n", nrow(X_train)))
  cat(sprintf("Test set size: %d\n", nrow(X_test)))
  
  return(list(
    X_train = X_train,
    y_train = y_train,
    X_test = X_test,
    y_test = y_test,
    train_data = train_data,
    test_data = test_data
  ))
}

#' Train Random Forest model
#' 
#' @param X_train Training features
#' @param y_train Training labels
#' @return Trained model
train_random_forest <- function(X_train, y_train) {
  cat("\nTraining Random Forest model...\n")
  
  model <- randomForest(
    x = X_train,
    y = y_train,
    ntree = 100,
    mtry = sqrt(ncol(X_train)),
    importance = TRUE,
    do.trace = FALSE
  )
  
  return(model)
}

#' Train XGBoost model
#' 
#' @param X_train Training features
#' @param y_train Training labels
#' @return Trained model
train_xgboost <- function(X_train, y_train) {
  cat("\nTraining XGBoost model...\n")
  
  # Convert to matrix and numeric labels
  train_matrix <- as.matrix(X_train)
  train_labels <- as.numeric(y_train) - 1  # Convert to 0/1
  
  # Train XGBoost
  model <- xgboost(
    data = train_matrix,
    label = train_labels,
    nrounds = 100,
    max_depth = 6,
    eta = 0.1,
    objective = "binary:logistic",
    eval_metric = "logloss",
    verbose = 0
  )
  
  return(model)
}

#' Train SVM model
#' 
#' @param X_train Training features
#' @param y_train Training labels
#' @return Trained model
train_svm <- function(X_train, y_train) {
  cat("\nTraining SVM model...\n")
  
  # Use a subset for SVM due to computational complexity
  sample_size <- min(5000, nrow(X_train))
  sample_idx <- sample(nrow(X_train), sample_size)
  X_train_sample <- X_train[sample_idx, ]
  y_train_sample <- y_train[sample_idx]
  
  model <- svm(
    x = X_train_sample,
    y = y_train_sample,
    kernel = "radial",
    probability = TRUE,
    scale = TRUE
  )
  
  return(model)
}

#' Train Neural Network model
#' 
#' @param train_data Training data with labels
#' @return Trained model
train_neural_network <- function(train_data) {
  cat("\nTraining Neural Network model...\n")
  
  # Prepare formula
  n_features <- ncol(train_data) - 1
  feature_names <- paste0("V", 1:n_features)
  formula_str <- paste("Label ~", paste(feature_names, collapse = " + "))
  formula <- as.formula(formula_str)
  
  # Scale data for neural network
  scaled_data <- train_data
  for (i in 1:n_features) {
    col_name <- feature_names[i]
    scaled_data[[col_name]] <- scale(scaled_data[[col_name]])
  }
  
  # Convert labels to numeric
  scaled_data$Label <- as.numeric(scaled_data$Label) - 1
  
  # Train neural network
  model <- neuralnet(
    formula = formula,
    data = scaled_data,
    hidden = c(64, 32),
    act.fct = "logistic",
    linear.output = FALSE,
    stepmax = 1e6
  )
  
  return(model)
}

#' Train Logistic Regression model
#' 
#' @param train_data Training data with labels
#' @return Trained model
train_logistic_regression <- function(train_data) {
  cat("\nTraining Logistic Regression model...\n")
  
  model <- glm(
    Label ~ .,
    data = train_data,
    family = binomial(link = "logit")
  )
  
  return(model)
}

#' Evaluate model performance
#' 
#' @param model Trained model
#' @param X_test Test features
#' @param y_test Test labels
#' @param model_type Type of model ("random_forest", "xgboost", "svm", "neural_network", "logistic")
#' @return List of performance metrics
evaluate_model <- function(model, X_test, y_test, model_type) {
  cat(sprintf("\nEvaluating %s model...\n", model_type))
  
  # Make predictions based on model type
  if (model_type == "random_forest") {
    y_pred <- predict(model, X_test, type = "response")
    y_pred_proba <- predict(model, X_test, type = "prob")[, 2]
  } else if (model_type == "xgboost") {
    test_matrix <- as.matrix(X_test)
    y_pred_proba <- predict(model, test_matrix)
    y_pred <- ifelse(y_pred_proba > 0.5, 1, 0)
    y_pred <- factor(y_pred, levels = c(0, 1))
  } else if (model_type == "svm") {
    y_pred <- predict(model, X_test)
    y_pred_proba <- attr(predict(model, X_test, probability = TRUE), "probabilities")[, 2]
  } else if (model_type == "neural_network") {
    # Scale test data
    test_scaled <- X_test
    for (i in 1:ncol(X_test)) {
      test_scaled[, i] <- scale(X_test[, i])
    }
    
    # Predict
    predictions <- compute(model, test_scaled)
    y_pred_proba <- predictions$net.result
    y_pred <- ifelse(y_pred_proba > 0.5, 1, 0)
    y_pred <- factor(y_pred, levels = c(0, 1))
  } else if (model_type == "logistic") {
    y_pred_proba <- predict(model, X_test, type = "response")
    y_pred <- ifelse(y_pred_proba > 0.5, 1, 0)
    y_pred <- factor(y_pred, levels = levels(y_test))
  }
  
  # Calculate metrics
  y_test_numeric <- as.numeric(y_test) - 1
  
  accuracy <- mean(y_pred == y_test)
  precision <- precision_score_func(y_test, y_pred)
  recall <- recall_score_func(y_test, y_pred)
  f1 <- f1_score_func(y_test, y_pred)
  
  # ROC AUC
  roc_obj <- roc(y_test_numeric, y_pred_proba, quiet = TRUE)
  roc_auc <- auc(roc_obj)
  
  # Confusion matrix
  cm <- confusionMatrix(y_pred, y_test)
  
  metrics <- list(
    accuracy = accuracy,
    precision = precision,
    recall = recall,
    f1 = f1,
    roc_auc = as.numeric(roc_auc),
    confusion_matrix = cm$table,
    y_pred = y_pred,
    y_pred_proba = y_pred_proba,
    roc_obj = roc_obj
  )
  
  return(metrics)
}

#' Calculate precision score
#' 
#' @param y_true True labels
#' @param y_pred Predicted labels
#' @return Precision score
precision_score_func <- function(y_true, y_pred) {
  cm <- table(y_true, y_pred)
  if (ncol(cm) == 2 && nrow(cm) == 2) {
    tp <- cm[2, 2]
    fp <- cm[1, 2]
    if (tp + fp == 0) return(0)
    return(tp / (tp + fp))
  }
  return(0)
}

#' Calculate recall score
#' 
#' @param y_true True labels
#' @param y_pred Predicted labels
#' @return Recall score
recall_score_func <- function(y_true, y_pred) {
  cm <- table(y_true, y_pred)
  if (ncol(cm) == 2 && nrow(cm) == 2) {
    tp <- cm[2, 2]
    fn <- cm[2, 1]
    if (tp + fn == 0) return(0)
    return(tp / (tp + fn))
  }
  return(0)
}

#' Calculate F1 score
#' 
#' @param y_true True labels
#' @param y_pred Predicted labels
#' @return F1 score
f1_score_func <- function(y_true, y_pred) {
  precision <- precision_score_func(y_true, y_pred)
  recall <- recall_score_func(y_true, y_pred)
  if (precision + recall == 0) return(0)
  return(2 * (precision * recall) / (precision + recall))
}

#' Plot confusion matrices
#' 
#' @param results List of results from all models
plot_confusion_matrices <- function(results) {
  models <- names(results)
  n_models <- length(models)
  
  plots <- list()
  for (i in 1:n_models) {
    model_name <- models[i]
    cm <- results[[model_name]]$confusion_matrix
    
    cm_df <- as.data.frame(cm)
    colnames(cm_df) <- c("Predicted", "Actual", "Freq")
    
    p <- ggplot(cm_df, aes(x = Predicted, y = Actual, fill = Freq)) +
      geom_tile() +
      geom_text(aes(label = Freq), color = "white", size = 5) +
      scale_fill_gradient(low = "lightblue", high = "darkblue") +
      labs(title = paste(toupper(model_name), "Confusion Matrix"),
           x = "Predicted Label", y = "True Label") +
      theme_minimal() +
      theme(plot.title = element_text(hjust = 0.5, size = 12, face = "bold"))
    
    plots[[i]] <- p
  }
  
  # Save plots
  if (n_models == 1) {
    ggsave("results/confusion_matrices.png", plots[[1]], width = 8, height = 6, dpi = 300)
  } else {
    # Combine plots using gridExtra
    library(gridExtra)
    combined_plot <- do.call(grid.arrange, c(plots, ncol = min(3, n_models)))
    ggsave("results/confusion_matrices.png", combined_plot, width = 5 * n_models, height = 4, dpi = 300)
  }
}

#' Plot ROC curves
#' 
#' @param results List of results from all models
plot_roc_curves <- function(results) {
  png("results/roc_curves.png", width = 10, height = 8, units = "in", res = 300)
  
  plot(0, 0, type = "n", xlim = c(0, 1), ylim = c(0, 1),
       xlab = "False Positive Rate", ylab = "True Positive Rate",
       main = "ROC Curves for All Models")
  
  colors <- rainbow(length(results))
  for (i in 1:length(results)) {
    model_name <- names(results)[i]
    roc_obj <- results[[model_name]]$roc_obj
    lines(1 - roc_obj$specificities, roc_obj$sensitivities,
          col = colors[i], lwd = 2, type = "l")
  }
  
  # Add diagonal line
  abline(0, 1, lty = 2, col = "gray")
  
  # Add legend
  legend("bottomright", legend = paste(toupper(names(results)), 
                                       sprintf("(AUC = %.3f)", 
                                               sapply(results, function(x) x$roc_auc))),
         col = colors, lwd = 2)
  
  grid()
  dev.off()
}

#' Save models and results
#' 
#' @param models List of trained models
#' @param results List of results
save_models_and_results <- function(models, results) {
  dir.create("models", showWarnings = FALSE)
  dir.create("results", showWarnings = FALSE)
  
  # Save models
  for (model_name in names(models)) {
    if (model_name == "xgboost") {
      xgb.save(models[[model_name]], sprintf("models/%s_model.model", model_name))
    } else if (model_name == "neural_network") {
      saveRDS(models[[model_name]], sprintf("models/%s_model.rds", model_name))
    } else {
      saveRDS(models[[model_name]], sprintf("models/%s_model.rds", model_name))
    }
  }
  
  # Save results summary
  results_summary <- data.frame(
    Model = names(results),
    Accuracy = sapply(results, function(x) x$accuracy),
    Precision = sapply(results, function(x) x$precision),
    Recall = sapply(results, function(x) x$recall),
    F1_Score = sapply(results, function(x) x$f1),
    ROC_AUC = sapply(results, function(x) x$roc_auc)
  )
  
  write.csv(results_summary, "results/model_comparison.csv", row.names = FALSE)
  
  # Save detailed results
  saveRDS(results, "results/detailed_results.rds")
  
  cat("\nModels and results saved!\n")
}

#' Main function to run the TF binding prediction pipeline
main <- function() {
  cat("=== Transcription Factor Binding Prediction Pipeline ===\n\n")
  
  # Create directories
  dir.create("results", showWarnings = FALSE)
  dir.create("models", showWarnings = FALSE)
  
  # Load and preprocess data
  data_list <- load_and_preprocess_data("data/genomics_data.csv")
  
  # Train models
  models <- list()
  models$random_forest <- train_random_forest(data_list$X_train, data_list$y_train)
  models$xgboost <- train_xgboost(data_list$X_train, data_list$y_train)
  models$svm <- train_svm(data_list$X_train, data_list$y_train)
  models$neural_network <- train_neural_network(data_list$train_data)
  models$logistic <- train_logistic_regression(data_list$train_data)
  
  # Evaluate models
  results <- list()
  results$random_forest <- evaluate_model(models$random_forest, data_list$X_test, 
                                          data_list$y_test, "random_forest")
  results$xgboost <- evaluate_model(models$xgboost, data_list$X_test, 
                                    data_list$y_test, "xgboost")
  results$svm <- evaluate_model(models$svm, data_list$X_test, 
                                data_list$y_test, "svm")
  results$neural_network <- evaluate_model(models$neural_network, data_list$X_test, 
                                           data_list$y_test, "neural_network")
  results$logistic <- evaluate_model(models$logistic, data_list$X_test, 
                                     data_list$y_test, "logistic")
  
  # Print results
  cat("\n=== Model Performance Summary ===\n")
  for (model_name in names(results)) {
    cat(sprintf("\n%s:\n", toupper(model_name)))
    cat(sprintf("  Accuracy:  %.4f\n", results[[model_name]]$accuracy))
    cat(sprintf("  Precision: %.4f\n", results[[model_name]]$precision))
    cat(sprintf("  Recall:    %.4f\n", results[[model_name]]$recall))
    cat(sprintf("  F1-Score:  %.4f\n", results[[model_name]]$f1))
    cat(sprintf("  ROC-AUC:   %.4f\n", results[[model_name]]$roc_auc))
  }
  
  # Plot results
  plot_confusion_matrices(results)
  plot_roc_curves(results)
  
  # Save models and results
  save_models_and_results(models, results)
  
  # Stop cluster
  stopCluster(cl)
  
  cat("\nPipeline completed successfully!\n")
  cat("Results saved in 'results/' directory\n")
  cat("Models saved in 'models/' directory\n")
}

# Run main function
if (!interactive()) {
  main()
}

